File: golf_ball_detector.py
import cv2
import time
from dataclasses import dataclass
from ultralytics import YOLO
import numpy as np

# Camera parameters and real world constants
ACTUAL_BALL_RADIUS = 2.135  # centimeters

FOCAL_LENGTH = 800.0        # pixels - approximate webcam focal length

# ArUco parameters
MARKER_SIZE = 5.0  # centimeters, length of one side of the marker


@dataclass
class BallMeasurement:
    """Stores a single measurement of the ball."""
    timestamp: float
    cx: float
    cy: float
    radius_px: float
    distance: float


def measure_ball(box) -> BallMeasurement:
    """Convert a YOLO bounding box to a BallMeasurement."""
    x1, y1, x2, y2 = box.xyxy[0]
    w = float(x2 - x1)
    h = float(y2 - y1)
    radius_px = (w + h) / 4.0  # average half width and half height
    cx = float(x1 + x2) / 2.0
    cy = float(y1 + y2) / 2.0
    distance = FOCAL_LENGTH * ACTUAL_BALL_RADIUS / radius_px
    return BallMeasurement(time.time(), cx, cy, radius_px, distance)



def main():
    model = YOLO('golf_ball_detector.onnx')


    # Scan for a working camera index
    def find_working_camera(max_index=5):
        for idx in range(max_index):
            for backend in [cv2.CAP_DSHOW, cv2.CAP_MSMF, None]:
                if backend is not None:
                    cap = cv2.VideoCapture(idx, backend)
                else:
                    cap = cv2.VideoCapture(idx)
                if cap.isOpened():
                    # Try to grab a frame
                    cap.set(cv2.CAP_PROP_FRAME_WIDTH, 1280)
                    cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 720)
                    for _ in range(5):
                        ret, frame = cap.read()
                        if ret and frame is not None and frame.mean() > 5:
                            print(f"Camera found at index {idx} with backend {backend if backend is not None else 'default'}.")
                            return cap
                    cap.release()
        print("No working camera found. Tried all indices and backends.")
        return None

    cap = find_working_camera(5)
    if cap is None or not cap.isOpened():
        print('Could not open any webcam. Try closing other apps that use the camera, check permissions, or try a different camera.')
        return


    # Test: Try to grab a frame and check if it's black, fallback to lower resolution if needed
    ret, frame = cap.read()
    if not ret or frame is None:
        print('Failed to grab initial frame from webcam.')
        cap.release()
        cv2.destroyAllWindows()
        return
    if frame.mean() < 5:
        print('First frame is all black, trying lower resolution (640x480)...')
        cap.set(cv2.CAP_PROP_FRAME_WIDTH, 640)
        cap.set(cv2.CAP_PROP_FRAME_HEIGHT, 480)
        for _ in range(5):
            ret, frame = cap.read()
            if ret and frame is not None and frame.mean() > 5:
                print('Successfully got a non-black frame at 640x480.')
                break
        else:
            print('Still getting black frames. Please check your camera index, close other apps, or try a different webcam.')
            cap.release()
            cv2.destroyAllWindows()
            return

    # Show the first frame for confirmation
    cv2.imshow('Webcam Test Frame', frame)
    print(f'Frame shape: {frame.shape}, mean pixel value: {frame.mean():.2f}')
    cv2.waitKey(1000)  # Show for 1 second
    cv2.destroyWindow('Webcam Test Frame')

    # Setup ArUco detection
    h, w = frame.shape[:2]
    camera_matrix = np.array(
        [[FOCAL_LENGTH, 0, w / 2.0], [0, FOCAL_LENGTH, h / 2.0], [0, 0, 1]],
        dtype=np.float32,
    )
    dist_coeffs = np.zeros((5, 1), dtype=np.float32)
    aruco_dict = cv2.aruco.getPredefinedDictionary(cv2.aruco.DICT_4X4_50)
    if hasattr(cv2.aruco, "DetectorParameters_create"):
        aruco_params = cv2.aruco.DetectorParameters_create()
    else:  # fallback for newer OpenCV where constructor is used directly
        aruco_params = cv2.aruco.DetectorParameters()

    # Some OpenCV versions expose ArUco detection via the ArucoDetector class
    if hasattr(cv2.aruco, "ArucoDetector"):
        aruco_detector = cv2.aruco.ArucoDetector(aruco_dict, aruco_params)
    else:
        aruco_detector = None

    # Main detection loop
    while True:
        ret, frame = cap.read()
        if not ret or frame is None:
            print('Failed to grab frame from webcam.')
            break

        gray = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
        if aruco_detector is not None:
            corners, ids, _ = aruco_detector.detectMarkers(gray)
        else:
            corners, ids, _ = cv2.aruco.detectMarkers(gray, aruco_dict, parameters=aruco_params)

        results = model(frame)
        annotated_frame = frame.copy()

        sticker_positions = []
        if ids is not None and len(ids) > 0:
            cv2.aruco.drawDetectedMarkers(annotated_frame, corners, ids)
            rvecs, tvecs, _ = cv2.aruco.estimatePoseSingleMarkers(
                corners, MARKER_SIZE, camera_matrix, dist_coeffs
            )
            for i, tvec in enumerate(tvecs):
                cv2.aruco.drawAxis(
                    annotated_frame, camera_matrix, dist_coeffs, rvecs[i], tvec, MARKER_SIZE / 2
                )
                sticker_positions.append((tvec[0][0], tvec[0][1], tvec[0][2] - 30.0))

        if results and len(results[0].boxes) > 0:
            boxes = results[0].boxes
            best_idx = boxes.conf.argmax()
            meas = measure_ball(boxes[best_idx])

            annotated_frame = results[0].plot()

            # Ball position relative to 30 cm origin
            bx = (meas.cx - w / 2.0) * meas.distance / FOCAL_LENGTH
            by = (meas.cy - h / 2.0) * meas.distance / FOCAL_LENGTH
            bz = meas.distance - 30.0
            ball_pos = (bx, by, bz)

            info = f"Ball:{[round(v,2) for v in ball_pos]}"
            for idx, pos in enumerate(sticker_positions):
                info += f" Sticker{idx+1}:{[round(v,2) for v in pos]}"
            cv2.putText(
                annotated_frame,
                info,
                (10, 30),
                cv2.FONT_HERSHEY_SIMPLEX,
                0.6,
                (0, 255, 0),
                2,
            )
            print(info)
        else:
            annotated_frame = frame

        cv2.imshow('Webcam Golf Ball Detection', annotated_frame)

        if cv2.waitKey(1) & 0xFF == ord('q'):
            break

    cap.release()
    cv2.destroyAllWindows()

if __name__ == '__main__':
    main()

File: golf_ball_detector.go
package main

import (
	"fmt"
	"image"
	"image/color"
	"time"

	ort "github.com/microsoft/onnxruntime-go"
	"gocv.io/x/gocv"
)

const (
	// Real golf ball radius in inches
	actualBallRadius = 2.135
	// Approximate focal length in pixels
	focalLength = 800.0
)

// BallMeasurement stores a detection converted to camera space.
type BallMeasurement struct {
	Timestamp float64
	CX        float64
	CY        float64
	RadiusPX  float64
	Distance  float64
}

// measureBall converts a YOLO box [x1,y1,x2,y2] to a measurement.
func measureBall(box [4]float32) BallMeasurement {
	x1, y1 := float64(box[0]), float64(box[1])
	x2, y2 := float64(box[2]), float64(box[3])
	w := x2 - x1
	h := y2 - y1
	radius := (w + h) / 4.0
	cx := (x1 + x2) / 2.0
	cy := (y1 + y2) / 2.0
	dist := focalLength * actualBallRadius / radius
	return BallMeasurement{
		Timestamp: float64(time.Now().UnixNano()) / 1e9,
		CX:        cx,
		CY:        cy,
		RadiusPX:  radius,
		Distance:  dist,
	}
}

// computeSpeed returns the velocity of the ball in camera space.
func computeSpeed(curr, prev BallMeasurement) (float64, float64, float64) {
	dt := curr.Timestamp - prev.Timestamp
	if dt <= 0 {
		return 0, 0, 0
	}
	avgZ := (curr.Distance + prev.Distance) / 2
	vx := ((curr.CX - prev.CX) / focalLength) * avgZ / dt
	vy := ((curr.CY - prev.CY) / focalLength) * avgZ / dt
	vz := (curr.Distance - prev.Distance) / dt
	return vx, vy, vz
}

func main() {
	// Open default webcam
	cam, err := gocv.OpenVideoCapture(0)
	if err != nil {
		fmt.Printf("failed to open camera: %v\n", err)
		return
	}
	defer cam.Close()

	win := gocv.NewWindow("Webcam Golf Ball Detection")
	defer win.Close()

	// Load ONNX model using onnxruntime
	session, err := ort.NewSession("golf_ball_detector.onnx")
	if err != nil {
		fmt.Printf("failed to load model: %v\n", err)
		return
	}
	defer session.Close()

	img := gocv.NewMat()
	defer img.Close()

	var prev *BallMeasurement

	for {
		if ok := cam.Read(&img); !ok || img.Empty() {
			continue
		}

		blob := gocv.BlobFromImage(img, 1.0/255.0, image.Pt(640, 640), gocv.NewScalar(0, 0, 0, 0), true, false)
		defer blob.Close()

		if err := session.SetInputTensor("images", blob); err != nil {
			fmt.Println("failed to set input tensor:", err)
			break
		}
		out, err := session.Run()
		if err != nil || len(out) == 0 {
			fmt.Println("inference error:", err)
			break
		}

		data, _ := out[0].Value().([]float32)
		// Each detection is [cx, cy, w, h, conf, class]
		var bestConf float32
		var bestBox [4]float32
		for i := 0; i+5 < len(data); i += 6 {
			conf := data[i+4]
			if conf > bestConf {
				cx := data[i]
				cy := data[i+1]
				w := data[i+2]
				h := data[i+3]
				bestBox = [4]float32{cx - w/2, cy - h/2, cx + w/2, cy + h/2}
				bestConf = conf
			}
		}

		if bestConf > 0 {
			meas := measureBall(bestBox)
			vx, vy, vz := 0.0, 0.0, 0.0
			if prev != nil {
				vx, vy, vz = computeSpeed(meas, *prev)
			}
			prev = &meas

			rect := image.Rect(int(bestBox[0]), int(bestBox[1]), int(bestBox[2]), int(bestBox[3]))
			gocv.Rectangle(&img, rect, color.RGBA{0, 255, 0, 0}, 2)
			info := fmt.Sprintf("Dist:%.2f in Vx:%.2f Vy:%.2f Vz:%.2f", meas.Distance, vx, vy, vz)
			gocv.PutText(&img, info, image.Pt(10, 30), gocv.FontHersheySimplex, 0.7, color.RGBA{0, 255, 0, 0}, 2)
		}

		win.IMShow(img)
		if win.WaitKey(1) == int('q') {
			break
		}
	}
}
